src/components/text.js

```diff
@@ -0,0 +1,169 @@
+        className: '',
+        fgColor: '#FFF',
+        round: false,
+        size: 100,
+        textSizeRatio: 3,
+        textMarginRatio: .15,
+        unstyled: false
+    }
+
+    componentDidMount() {
+        this._mounted = true;
+        this._scaleTextNode(this._node);
+    }
+
+    componentWillUnmount() {
+        this._mounted = false;
+    }
+
+    _scaleTextNode = (node, retryTTL = 16) => {
+        const {
+            unstyled,
+            textSizeRatio,
+            textMarginRatio,
+            avatar
+        } = this.props;
+
+        this._node = node;
+
+        if (!node || unstyled || avatar.src || !this._mounted)
+            return;
+
+        const spanNode = node.parentNode;
+        const tableNode = spanNode.parentNode;
+
+        const {
+            width: containerWidth,
+            height: containerHeight
+        } = spanNode.getBoundingClientRect();
+
+        // Whenever the avatar element is not visible due to some CSS
+        // (such as display: none) on any parent component we will check
+        // whether the component has become visible.
+        //
+        // The time between checks grows up to half a second in an attempt
+        // to reduce flicker / performance issues.
+        if (containerWidth == 0 && containerHeight == 0) {
+            const ttl = Math.min(retryTTL * 1.5, 500);
+            setGroupedTimeout(() => this._scaleTextNode(node, ttl), ttl);
+            return;
+        }
+
+        // If the tableNode (outer-container) does not have its fontSize set yet,
+        // we'll set it according to "textSizeRatio"
+        if (!tableNode.style.fontSize) {
+            const baseFontSize = containerHeight / textSizeRatio;
+            tableNode.style.fontSize = `${baseFontSize}px`;
+        }
+
+        // Reset font-size such that scaling works correctly (#133)
+        spanNode.style.fontSize = null;
+
+        // Measure the actual width of the text after setting the container size
+        const { width: textWidth } = node.getBoundingClientRect();
+
+        if (textWidth < 0)
+            return;
+
+        // Calculate the maximum width for the text based on "textMarginRatio"
+        const maxTextWidth = containerWidth * (1 - (2 * textMarginRatio));
+
+        // If the text is too wide, scale it down by (maxWidth / actualWidth)
+        if (textWidth > maxTextWidth)
+            spanNode.style.fontSize = `calc(1em * ${maxTextWidth / textWidth})`;
+    }
+
+    render() {
+        const {
+            className,
+            round,
+            unstyled,
+            title,
+            name,
+            value,
+            avatar
+        } = this.props;
+
+        const size = parseSize(this.props.size);
+
+        const initialsStyle = unstyled ? null : {
+            width: size.str,
+            height: size.str,
+            lineHeight: 'initial',
+            textAlign: 'center',
+            color: this.props.fgColor,
+            background: avatar.color,
```

Comment:
This breaks the last Avatar on the demo page:
`<CustomAvatar githubHandle="JorgenEvens" color="#E88554" />`
Where `CustomAvatar` is configured with these 2 sources: `GravatarSource`, `ValueSource`.

I don't know if the `githubHandle` prop is an error or if this is intended to show that, when you provide incompatible props with regard to the sources, it fallbacks to the `color` prop.
This is no longer the case here since `avatar` will have all its fields to null.

Reply:
@JorgenEvens LGTM apart from that.
If you rebase on master I can also check if DataProvider works in React Native